waspc/test/Analyzer/TypeChecker/InternalTest.hs

```diff
@@ -0,0 +1,202 @@
+module Analyzer.TypeChecker.InternalTest where
+
+import qualified Analyzer.Parser as P
+import Analyzer.Type
+import Analyzer.TypeChecker.AST
+import Analyzer.TypeChecker.Internal
+import Analyzer.TypeChecker.Monad (Bindings, runT, runTWithBound)
+import Analyzer.TypeChecker.TypeError
+import qualified Analyzer.TypeDefinitions as TD
+import qualified Data.HashMap.Strict as H
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import Test.Tasty.Hspec
+import Test.Tasty.QuickCheck
+
+-- TODO:
+-- Having Arbitrary instances for Types and TypedAST would allow much more thorough testing
+-- of the type checker, and is probably worth the effort to figure out how to make instances
+-- given how complex the code is, how many corner cases there are, and how well-defined the
+-- required properties of the type-checking functions are.
+
+chooseType :: Gen Type
+chooseType =
+  elements
+    [ StringType,
+      NumberType,
+      BoolType,
+      ExtImportType,
+      QuoterType "json",
+      QuoterType "psl"
+    ]
+
+checkExpr' :: Bindings -> P.Expr -> Either TypeError TypedExpr
+checkExpr' bindings expr = runTWithBound bindings TD.empty $ checkExpr expr
+
+spec_Parser :: Spec
+spec_Parser = do
+  describe "Analyzer.TypeChecker.Internal" $ do
+    describe "unify" $ do
+      it "Doesn't affect 2 expressions of the same type" $ do
+        property $ \(a, b) ->
+          let initial = IntegerLiteral a :| [DoubleLiteral b]
+              actual = unify initial
+           in actual == Right (initial, NumberType)
+      it "Unifies two same-typed dictionaries to their original type" $ do
+        let typ = DictType $ H.fromList [("a", DictRequired BoolType), ("b", DictOptional NumberType)]
```

Comment:
Hm it would be cool if we could just parse these directly from the code, maybe even during compile time! So instead of writing AST we would be writing Wasp code for these expressions, in tests. But I guess our parser can't do that at the moment?

Reply:
Yes, wasp quasiquotes!

That is reasonable, avoiding bugs in parser causing bugs in typechecker. Hmmm. But on the other hand, we do have tests for parser and those will detect errors if they happen. I think from the TDD and Clean Code perspective, so following the wisdom of books, you are right, it is better to have this kind of separation. However on the practical side, I would be ok with sacrificing that separation in this specific case, if it means TypeChecker tests would be easier to read and maintain, which is an important benefit -> it encourages devs to be less afraid of modifying tests like these ones that have a lot of code.

I am not saying we have to do this now, it is probably too much a digression, but maybe we could leave a TODO for the future. What is your thought regarding implementation -> I am guessing our parser currently doesn't support parsing only snippets of Wasp code?

Reply:
Great, let's leave a TODO!